### PR TITLE
#492 Edit categories redesign

### LINKED ISSUE #492: Update Category List UI
**Acceptance Criteria**:

1. The space for each category name now spans to fill the width of the screen (currently: the user has to land on the arrow on open each category)
2.  New design reflects less negative space and allows user more room to select each category

**Design**:

[Ipad Landscape](https://www.figma.com/file/mNrwUygVhTmuWnKNHHkenR/Vocable---Current?node-id=461%3A0)
[Ipad Portrait](https://www.figma.com/file/mNrwUygVhTmuWnKNHHkenR/Vocable---Current?node-id=461%3A901)
[Iphone Landscape](https://www.figma.com/file/mNrwUygVhTmuWnKNHHkenR/Vocable---Current?node-id=461%3A1214)
[Iphone Portrait](https://www.figma.com/file/mNrwUygVhTmuWnKNHHkenR/Vocable---Current?node-id=461%3A1293)

### PR DESCRIPTION
**Pull request by Clstroud**
closes #492 

## Description of Work

* Fixes a bug with the carousel grid layout that improves row/column math handling
* Builds on `VocableListCellConfiguration` for use in `EditCategoriesViewController`
* Implements `VocableListCell` in `EditCategoriesViewController`
 
## Notes to Test 
* Rotation and the resulting scroll position is just generally wonky right now and I don't like it. Doesn't appear to be anything introduced by this ticket, but it's driving me bananas while double checking this work's functionality

## Screenshots 
| Device      | Portrait | Landscape |
| --- | --- | --- |
| iPad      | Simulator Screen Shot - iPad Air (4th generation) - 2022-03-25 at 14 17 17       | Simulator Screen Shot - iPad Air (4th generation) - 2022-03-25 at 14 17 21 |
| iPhone  | Simulator Screen Shot - iPhone 12 Pro - 2022-03-25 at 14 14 57 | Simulator Screen Shot - iPhone 12 Pro - 2022-03-25 at 14 14 43       |

### CODE REVIEW

**Comment by SashankPatel** [comment]
Test Notes 3/28
- Categories layout match figma on iPhone (testing on iPhone 8 simulator).
- Tested up/down arrows and navigating into category
- Tested orientation.
- Verified layout matches figma on iPad.

OBSERVATION: As noted, I also noticed when launching the app there is a flicker of categories scrolling.  Also carousel through the categories is choppy/slower.

Looks good from test.

File: .swiftlint.yml [hunk 1]
```diff
@@ -38,4 +38,6 @@ file_length:
 cyclomatic_complexity:
     warning: 15
     error: 25
+nesting:
+    type_level: 5
```
Comment (moyerr) [inline .swiftlint.yml:42]: 💯

File: Vocable/Common/VocableListCell/VocableListContentConfiguration.swift [hunk 2]
```diff
@@ -10,36 +10,102 @@ import UIKit
...
         VocableListCellContentView(configuration: self)
     }
 
     func updated(for state: UIConfigurationState) -> VocableListContentConfiguration {
-        return self
+
+        var updatedSelf = self
+        if let handler = traitCollectionChangeHandler {
+            handler(state.traitCollection, &updatedSelf)
+        }
```
Comment (moyerr) [inline Vocable/Common/VocableListCell/VocableListContentConfiguration.swift]:
```suggestion
        traitCollectionChangeHandler?(state.traitCollection, &updatedSelf)
```

File: Vocable/Features/Settings/EditCategories/EditCategoriesViewController.swift [hunk 3]
```diff
@@ -121,131 +144,118 @@ final class EditCategoriesViewController: PagingCarouselViewController, NSFetche
...
 
-        let rowOrdinal = indexPath.row + 1
-        let formattedRowOrdinal = EditCategoriesViewController.rowIndexFormatter.string(from: NSNumber(value: rowOrdinal)) ?? "\(rowOrdinal)"
-        let visibleTitleString = NSMutableAttributedString(string: "\(formattedRowOrdinal). \(category.name ?? "")")
-        cell.showCategoryDetailButton.isEnabled = true
-        cell.setup(title: visibleTitleString)
+        if #available(iOS 15, *) {
+            print("Use snapshot-based reconfiguring instead")
+            return
+        }
```
Comment (moyerr) [inline Vocable/Features/Settings/EditCategories/EditCategoriesViewController.swift]: Could we mark `updateVisibleCellConfigurations()` as unavailable in iOS 15 instead?

**Review by moyerr — APPROVED** [review]
Left some minor comments, but nothing blocking. Looks great 🏆

File: Vocable/Features/Settings/EditCategories/EditCategoriesViewController.swift [hunk 4]
```diff
@@ -81,31 +72,63 @@ final class EditCategoriesViewController: PagingCarouselViewController, NSFetche
     }
 
     private func setupCollectionView() {
-        collectionView.register(UINib(nibName: "EditCategoriesDefaultCollectionViewCell", bundle: nil), forCellWithReuseIdentifier: EditCategoriesDefaultCollectionViewCell.reuseIdentifier)
-        collectionView.register(UINib(nibName: "EditCategoriesCompactCollectionViewCell", bundle: nil), forCellWithReuseIdentifier: "EditCategoriesCompactCollectionViewCell")
         collectionView.backgroundColor = .collectionViewBackgroundColor
+        cellRegistration = UICollectionView.CellRegistration<VocableListCell, Category>(handler: updateContentConfiguration)
```
Comment (moyerr) [inline Vocable/Features/Settings/EditCategories/EditCategoriesViewController.swift]: Oh I just noticed this one while working on my branch for #485. Passing the method like this actually captures a strong reference of `self` and will result in a retain cycle.

Comment (Clstroud) [reply Vocable/Features/Settings/EditCategories/EditCategoriesViewController.swift]: Crap, you're right. I can fix that! Good call.

**Comment by SashankPatel** [comment]
Test update:
Verified on iPhone 8 iOS 14 simulator and iPad 9th gen iOS 15 simulator.
- Category name is now displayed when adding a category.
     - App does not crash when selecting the category.
- Categories can be deleted and repositioned.

Observation: When in the Settings view.  If you are on page 3 of 3 and a new category is added, instead of going to page 4 of 4, it would jump to 1 of 4.  When deleting the category from page 4 of 4, instead of going to 3 of 3, it would jump to 2 of 3.  This happened consistently, but after restarting the app it wasn't happening.  There is an existing bug that is similar to this #495 - CC @rhsalas3

**Pull request merged**